0.23.0 fails to build on Alpine

I'm currently seeing [this build error](https://github.com/acj/aport-builds/actions/runs/14305129235/job/40087271141) on both 0.23.0 and the tip of master (currently d5653cf):

```
ninja: job failed: /usr/bin/g++ -DKERNEL_HEADERS_DIR=\"\" -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/util/CMakeFiles/util.dir/paths.cpp.o -MF src/util/CMakeFiles/util.dir/paths.cpp.o.d -o src/util/CMakeFiles/util.dir/paths.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/log.h:11,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp:10:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
```

There are a few old issues referring to location.hh as an ephemeral file that's generated during the build. Is that still the case? If so, maybe there's an order of operations issue?

It looks like a few references to location.hh have been removed in recent commits, leaving only [these](https://github.com/search?q=repo%3Abpftrace%2Fbpftrace%20%22location.hh%22&type=code) in location.h.

Thanks. Applying that patch to master, I still get

```
[53/147] Building CXX object src/util/CMakeFiles/util.dir/paths.cpp.o
ninja: job failed: /usr/bin/g++ -DKERNEL_HEADERS_DIR=\"\" -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/util/CMakeFiles/util.dir/system.cpp.o -MF src/util/CMakeFiles/util.dir/system.cpp.o.d -o src/util/CMakeFiles/util.dir/system.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/system.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/log.h:11,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/system.cpp:8:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
ninja: job failed: /usr/bin/g++ -DKERNEL_HEADERS_DIR=\"\" -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/util/CMakeFiles/util.dir/paths.cpp.o -MF src/util/CMakeFiles/util.dir/paths.cpp.o.d -o src/util/CMakeFiles/util.dir/paths.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/log.h:11,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp:10:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
```

I tried adding `add_dependencies(compiler_core parser)` in `src/CMakeLists.txt` and `src/ast/CMakeLists.txt` but that didn't seem to change anything. Where is the correct place for it?

I don't have a simple repro to offer for this, unfortunately, but I'm happy to test fixes.

I got this on the first build:

```
ninja: job failed: /usr/bin/g++ -DKERNEL_HEADERS_DIR=\"\" -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/util/CMakeFiles/util.dir/io.cpp.o -MF src/util/CMakeFiles/util.dir/io.cpp.o.d -o src/util/CMakeFiles/util.dir/io.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp: In member function 'void bpftrace::util::StdioSilencer::silence()':
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp:24:23: error: 'dup' was not declared in this scope
   24 |     if ((old_stdio_ = dup(fd)) < 0)
      |                       ^~~
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp:31:9: error: 'dup2' was not declared in this scope
   31 |     if (dup2(new_stdio, fd) < 0)
      |         ^~~~
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp:34:5: error: 'close' was not declared in this scope; did you mean 'pclose'?
   34 |     close(new_stdio);
      |     ^~~~~
      |     pclose
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp: In destructor 'bpftrace::util::StdioSilencer::~StdioSilencer()':
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp:58:9: error: 'dup2' was not declared in this scope
   58 |     if (dup2(old_stdio_, fd) < 0)
      |         ^~~~
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/io.cpp:60:5: error: 'close' was not declared in this scope; did you mean 'pclose'?
   60 |     close(old_stdio_);
      |     ^~~~~
      |     pclose
ninja: job failed: /usr/bin/g++ -DKERNEL_HEADERS_DIR=\"\" -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/util/CMakeFiles/util.dir/paths.cpp.o -MF src/util/CMakeFiles/util.dir/paths.cpp.o.d -o src/util/CMakeFiles/util.dir/paths.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp: In lambda function:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/util/paths.cpp:90:7: error: '::close' has not been declared; did you mean 'pclose'?
   90 |     ::close(fd);
      |       ^~~~~
      |       pclose
ninja: subcommands failed
```

Added these to address it:

```diff
diff --git src/util/io.cpp src/util/io.cpp
index 721c1eff..d48a52c1 100644
--- a/src/util/io.cpp
+++ b/src/util/io.cpp
@@ -1,6 +1,7 @@
 #include <cstring>
 #include <fcntl.h>
 #include <fstream>
+#include <unistd.h>

 #include "log.h"
 #include "util/exceptions.h"
diff --git src/util/paths.cpp src/util/paths.cpp
index 8af5e1b4..3b662a19 100644
--- a/src/util/paths.cpp
+++ b/src/util/paths.cpp
@@ -6,6 +6,7 @@
 #include <glob.h>
 #include <iostream>
 #include <regex>
+#include <unistd.h>

 #include "log.h"
 #include "scopeguard.h"
```

Then this error, same as before (or at least similar:

```
ninja: job failed: /usr/bin/g++  -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/CMakeFiles/compiler_core.dir/types.cpp.o -MF src/CMakeFiles/compiler_core.dir/types.cpp.o.d -o src/CMakeFiles/compiler_core.dir/types.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/types.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/diagnostic.h:10,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/ast.h:7,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/struct.h:8,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/types.cpp:7:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
ninja: job failed: /usr/bin/g++  -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/CMakeFiles/compiler_core.dir/struct.cpp.o -MF src/CMakeFiles/compiler_core.dir/struct.cpp.o.d -o src/CMakeFiles/compiler_core.dir/struct.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/struct.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/diagnostic.h:10,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/ast.h:7,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/struct.h:8,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/struct.cpp:6:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
ninja: job failed: /usr/bin/g++  -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src -I/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/build -isystem /usr/lib/llvm18/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_ENABLE_HARDENED_MODE=1 -fno-plt -Os -DNDEBUG -std=c++20   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Werror=missing-field-initializers -Werror=undef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wunreachable-code -Wdisabled-optimization -fdiagnostics-color=always -MD -MT src/CMakeFiles/compiler_core.dir/functions.cpp.o -MF src/CMakeFiles/compiler_core.dir/functions.cpp.o.d -o src/CMakeFiles/compiler_core.dir/functions.cpp.o -c /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/functions.cpp
In file included from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/diagnostic.h:10,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/ast.h:7,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/functions.h:9,
                 from /home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/functions.cpp:1:
/home/builder/package/src/bpftrace-d5653cf8c1e2519cbbf37978d17f855b6323e438/src/ast/location.h:9:10: fatal error: location.hh: No such file or directory
    9 | #include "location.hh"
      |          ^~~~~~~~~~~~~
compilation terminated.
```

Added this back in:

```diff
diff --git src/CMakeLists.txt src/CMakeLists.txt
index ced47076..ad877840 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -10,4 +10,5 @@ add_library(compiler_core STATIC
   struct.cpp
   types.cpp
 )
+add_dependencies(compiler_core parser)

```

And then it built successfully.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -39,6 +39,8 @@ and this project adheres to
 #### Deprecated
 #### Removed
 #### Fixed
+- Fix build failures due to missing location.hh
+  - [#3987](https://github.com/bpftrace/bpftrace/pull/3987)
 #### Security
 #### Docs
 #### Tools

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -10,6 +10,7 @@ add_library(compiler_core STATIC
   struct.cpp
   types.cpp
 )
+add_dependencies(compiler_core parser)
 
 add_library(runtime STATIC
   attached_probe.cpp

--- a/src/ast/CMakeLists.txt
+++ b/src/ast/CMakeLists.txt
@@ -34,6 +34,7 @@ add_library(ast STATIC
   passes/recursion_check.cpp
 )
 
+add_dependencies(ast parser)
 target_compile_definitions(ast PRIVATE ${BPFTRACE_FLAGS})
 target_link_libraries(ast PRIVATE debugfs tracefs util)
 target_link_libraries(ast PUBLIC ast_defs arch compiler_core parser)

--- a/src/bpftrace.h
+++ b/src/bpftrace.h
@@ -15,7 +15,6 @@
 #include <vector>
 
 #include "ast/ast.h"
-#include "ast/location.h"
 #include "ast/pass_manager.h"
 #include "attached_probe.h"
 #include "bpfbytecode.h"

--- a/src/functions.h
+++ b/src/functions.h
@@ -7,7 +7,6 @@
 #include <vector>
 
 #include "ast/ast.h"
-#include "ast/location.h"
 #include "types.h"
 #include "util/hash.h"
 

--- a/src/log.h
+++ b/src/log.h
@@ -8,8 +8,6 @@
 #include <unordered_map>
 #include <vector>
 
-#include "ast/location.h"
-
 namespace bpftrace {
 
 namespace LogColor {

--- a/src/output.h
+++ b/src/output.h
@@ -4,7 +4,6 @@
 #include <map>
 #include <vector>
 
-#include "ast/location.h"
 #include "bpfmap.h"
 #include "required_resources.h"
 #include "types.h"

--- a/src/util/io.cpp
+++ b/src/util/io.cpp
@@ -1,6 +1,7 @@
 #include <cstring>
 #include <fcntl.h>
 #include <fstream>
+#include <unistd.h>
 
 #include "log.h"
 #include "util/exceptions.h"

--- a/src/util/paths.cpp
+++ b/src/util/paths.cpp
@@ -6,6 +6,7 @@
 #include <glob.h>
 #include <iostream>
 #include <regex>
+#include <unistd.h>
 
 #include "log.h"
 #include "scopeguard.h"

--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -1,5 +1,6 @@
 #include <algorithm>
 #include <array>
+#include <climits>
 #include <filesystem>
 #include <fstream>
 #include <map>

